website/events/api/v2/views.py

```diff
@@ -38,12 +47,118 @@ class EventDetailView(RetrieveAPIView):
     required_scopes = ["events:read"]
 
 
-class EventRegistrationListView(ListAPIView):
+class EventRegistrationsView(ListAPIView):
+    """Returns a list of registrations."""
 
     serializer_class = EventRegistrationSerializer
+    permission_classes = [
+        IsAuthenticatedOrTokenHasScopeForMethod,
+        DjangoModelPermissionsOrAnonReadOnly,
+    ]
+    required_scopes_per_method = {
+        "GET": ["events:read"],
+        "POST": ["events:register"],
+        "DELETE": ["events:register"],
+    }
+
+    def get_serializer_class(self):
+        if self.request.method.lower() == "post":
+            return EmptySerializer
+        return super().get_serializer_class()
 
     def get_queryset(self):
-        event = get_object_or_404(Event, pk=self.kwargs.get("pk"))
-        return EventRegistration.objects.filter(event=event.pk, date_cancelled=None)[
-            : event.max_participants
-        ]
+        return EventRegistration.objects.filter(
+            event=self.event.pk, date_cancelled=None
+        )[: self.event.max_participants]
+
+    def dispatch(self, request, *args, **kwargs):
+        self.event = get_object_or_404(Event, pk=self.kwargs.get("pk"))
+        return super().dispatch(request, *args, **kwargs)
+
+    def post(self, request, *args, **kwargs):
+        try:
+            registration = services.create_registration(request.member, self.event)
+            serializer = EventRegistrationSerializer(
+                instance=registration, context=self.get_serializer_context()
+            )
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        except RegistrationError as e:
+            raise PermissionDenied(detail=e) from e
+
+    def delete(self, request, *args, **kwargs):
+        try:
+            services.cancel_registration(request.member, self.event)
+            return Response(status=status.HTTP_204_NO_CONTENT)
+        except RegistrationError as e:
+            raise PermissionDenied(detail=e) from e
+
+
+class EventRegistrationDetailView(RetrieveAPIView):
+    """Returns details of an event registration."""
+
+    serializer_class = EventRegistrationSerializer
+    queryset = EventRegistration.objects.all()
+    permission_classes = [
+        IsAuthenticatedOrTokenHasScope,
+        DjangoModelPermissionsOrAnonReadOnly,
+    ]
+    required_scopes = ["events:read"]
+
+    def get_queryset(self):
+        return super().get_queryset().filter(event=self.kwargs["event_id"])
+
+
+class EventRegistrationFieldsView(APIView):
+    """Returns details of an event registration."""
+
+    permission_classes = [IsAuthenticatedOrTokenHasScopeForMethod]
```

Comment:
Shouldn't this also contain a DjangoModelPermissions permission class as well?

Reply:
This APIView has no model so that is impossible, but the services.py should already handle that.